// better-auth/dist/plugins/index.mjs
import "../dialect-BvcuosDD.mjs";
import "../get-model-name-D4DUV7S2.mjs";
import "../get-migration-C2PG4G2s.mjs";
import { n as HIDE_METADATA } from "../utils-DBbaShi0.mjs";
import "../crypto-BQxYXGGX.mjs";
import "../jwt-DzuIBp7t.mjs";
import "../misc-D8A8FGv2.mjs";
import "../schema-CSB7dlK0.mjs";
import "../get-request-ip-RYTz2Uba.mjs";
import "../json-_HMgPUVh.mjs";
import "../url-C4fqRUH1.mjs";
import "../api-CE_W4rUg.mjs";
import "../cookies-m6bKaTlj.mjs";
import "../session-BO-S7wVw.mjs";
import "../password-DBH2q5D6.mjs";
import "../parser-kwbb5q_C.mjs";
import { n as role, t as createAccessControl } from "../access-DXuPG0DH.mjs";
import "../access-Bue6UE97.mjs";
import "../has-permission-DN8I_Icv.mjs";
import "../access-Cr_p-YVH.mjs";
import "../permission-ouQCO-51.mjs";
import { t as twoFactorClient } from "../client-CZZjjkJj.mjs";
import "../plugin-helper-CA_Vvnl9.mjs";
import { t as admin } from "../admin-BWblrERx.mjs";
import { t as anonymous } from "../anonymous-B03YuGJE.mjs";
import { a as getMCPProtectedResourceMetadata, c as oAuthDiscoveryMetadata, d as lastLoginMethod, f as API_KEY_TABLE_NAME, h as defaultKeyHasher, i as oneTimeToken, l as oAuthProtectedResourceMetadata, m as apiKey, n as createAuthMiddleware, o as getMCPProviderMetadata, p as ERROR_CODES, r as optionsMiddleware, s as mcp, t as createAuthEndpoint, u as withMcpAuth } from "../plugins-CBWTGEDd.mjs";
import { t as bearer } from "../bearer-C42aRZlK.mjs";
import { t as captcha } from "../captcha-Dnh5mC7N.mjs";
import { t as customSession } from "../custom-session-D7ueXXp_.mjs";
import { n as deviceAuthorizationOptionsSchema, t as deviceAuthorization } from "../device-authorization-Dr11VVmd.mjs";
import { t as emailOTP } from "../email-otp-fvjP9Km7.mjs";
import { a as line, c as auth0, i as microsoftEntraId, n as slack, o as keycloak, r as okta, s as hubspot, t as genericOAuth } from "../generic-oauth-qo6urBZS.mjs";
import { t as haveIBeenPwned } from "../haveibeenpwned-1NCWtMXr.mjs";
import { a as generateExportedKeyPair, i as createJwk, n as verifyJWT, r as getJwtToken, t as jwt } from "../jwt-COIliG0M.mjs";
import { t as magicLink } from "../magic-link-U7vFkXXA.mjs";
import { n as getMetadata, r as oidcProvider, t as getClient } from "../oidc-provider-DiyW2arC.mjs";
import { t as multiSession } from "../multi-session-C9LCNQiw.mjs";
import { t as oAuthProxy } from "../oauth-proxy-CZ_6azrG.mjs";
import { t as oneTap } from "../one-tap-Djbp5wgR.mjs";
import { t as openAPI } from "../open-api-BDqC0qwL.mjs";
import { n as parseRoles, t as organization } from "../organization-CL4Xa8Vi.mjs";
import { t as phoneNumber } from "../phone-number-BnV6uyZR.mjs";
import { t as siwe } from "../siwe-DhIExJIr.mjs";
import { n as TWO_FACTOR_ERROR_CODES, t as twoFactor } from "../two-factor-awmDvK8-.mjs";
import { n as USERNAME_ERROR_CODES, t as username } from "../username-Drs37y_f.mjs";

export { API_KEY_TABLE_NAME, ERROR_CODES, HIDE_METADATA, TWO_FACTOR_ERROR_CODES, USERNAME_ERROR_CODES, admin, anonymous, apiKey, auth0, bearer, captcha, createAccessControl, createAuthEndpoint, createAuthMiddleware, createJwk, customSession, defaultKeyHasher, deviceAuthorization, deviceAuthorizationOptionsSchema, emailOTP, generateExportedKeyPair, genericOAuth, getClient, getJwtToken, getMCPProtectedResourceMetadata, getMCPProviderMetadata, getMetadata, haveIBeenPwned, hubspot, jwt, keycloak, lastLoginMethod, line, magicLink, mcp, microsoftEntraId, multiSession, oAuthDiscoveryMetadata, oAuthProtectedResourceMetadata, oAuthProxy, oidcProvider, okta, oneTap, oneTimeToken, openAPI, optionsMiddleware, organization, parseRoles, phoneNumber, role, siwe, slack, twoFactor, twoFactorClient, username, verifyJWT, withMcpAuth };
